src/api/integration/transactions_test.go

```diff
@@ -253,35 +288,314 @@ func TestLiveInjectTransactionEnableNetworking(t *testing.T) {
+				txn, _ := prepareTxnFunc(t, w.Entries[0].Address.String(), totalCoins, "1")
+				// Use a wrong private key to sign txn.In[0] and change txn.Sigs[0]
+				_, seckey := cipher.GenerateKeyPair()
+				h := cipher.AddSHA256(txn.InnerHash, txn.In[0])
+				txn.Sigs[0] = cipher.MustSignHash(h, seckey)
+
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Signature not valid for output being spent",
+		},
+		{
+			name: "insufficient coins",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), totalCoins, "1")
+				// Make output coins > input coins
+				txn.Out[0].Coins = txn.Out[0].Coins + 1
+				txn.InnerHash = txn.HashInner()
+				// Sign txn again as the inner hash is changed
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Insufficient coins",
+		},
+		{
+			name: "transaction may not destry coins",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), totalCoins, "1")
+				// Make output coins < input coins
+				txn.Out[0].Coins = txn.Out[0].Coins - 1
+				txn.InnerHash = txn.HashInner()
+				// Sign the txn again as the inner hash is changed
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Transactions may not destroy coins",
+		},
+		{
+			name: "insufficient coin hours",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), totalCoins, "1")
+				// Make up more output coin hours
+				txn.Out[0].Hours = txn.Out[0].Hours + 1e6
+				// Recalculate inner hash
+				txn.InnerHash = txn.HashInner()
+				// Sign the txn again
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Insufficient coin hours",
+		},
+		{
+			name: "invalid amount, too many decimal places",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				// Make a txn with txn.Out[0].Coins equal 1e3, as we have at least 2e6 coins
+				// so there will have at least two outputs.
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), 1e3, "1")
+				// Make txn.Out[0].Coins too many decimal places
+				txn.Out[0].Coins = 5e2
+				// Move the remaining 5e2 from the first output to the second output, so that
+				// we won't lose coins
+				txn.Out[1].Coins = txn.Out[1].Coins + 5e2
+				txn.InnerHash = txn.HashInner()
+
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates soft constraint: invalid amount, too many decimal places",
 		},
 	}
+	// TODO:
+	// The following test cases can not be added here, as we cannot use the big size
```

Comment:
Thats ok, the main thing we are testing in the integration test is that a 400 error occurs for any invalid txn injection, and also that the txn is really not added to the db.   Its ok if we don't test for every possible invalid txn type, we test that in unit tests